Removes import stage calculating dense nodes

tinwelint opened the pull request:
And instead only has one step reading relationship input. Introduced another step which links relationship chains together (link forward) before linking backwards. This approach seems faster than the previous approach.

The forked processor has seen a rewrite for better throughput. Forked processor is a step which processes batches sequentially, although it has potentially multiple internal processors, each processing only its own part of a batch e.g. based on node id. This is very useful for parallelizing processing of batches that uses unsynchronized, or non-trivially-synchronizable data structures.

tinwelint committed:
Removes explicit CalculateDenseNodeStage from importer

by instead importing relationships when going through the relationship input
w/o any linking. Then have one pass linking all relationships forwards before
linking them backwards. In most cases this should be faster than the previous
approach.

tinwelint committed:
Include page cache in used memory

tinwelint committed:
Faster forked processor step

by doing way less unpark and instead implementing a design where each forked
processor can walk the queue of batches individually. In the previous design
the act of unparking all the forked processors became the dominating operation
when number of assigned processors went up. Now profiling doesn't even show it
and throughput is better.

tinwelint commented:
Builds on #9273

burqen commented:
@tinwelint You got yourself some checkstyle errors.
`ForkedProcessorStepTest.java:29:1: Redundant import from the same package`

tinwelint committed:
Checkstyle

tinwelint committed:
Awaits current batches to complete before changing processor count

in ForkedProcessorStep, so that there may not be an order inconsistency.
The order inconsistency would come from there being a different
processor count and so the distribution of work would be different,
which means that a new particular path of a batch may be assigned to
a different processor than the one getting those parts, before the
processor change. This would introduce a race between the new and old
processor for that particular part.

Solution here is to drain the processing queue before applying the new
processor count.

tinwelint commented:
@burqen looks like green with an unrelated flaky test

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipLinkforwardStage.java:
```diff
@@ -0,0 +1,46 @@
+import static org.neo4j.unsafe.impl.batchimport.RecordIdIterator.forwards;
+import static org.neo4j.unsafe.impl.batchimport.staging.Step.ORDER_SEND_DOWNSTREAM;
+
+public class RelationshipLinkforwardStage extends Stage
```
Please add javadoc similar to the other stages, they help a lot.

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipStage.java:
```diff
@@ -19,84 +19,53 @@
- * are appended to the end, only "next" pointers can be provided in new records in the face of the sequential-only
- * I/O restriction and {@link RelationshipGroupRecord} chains must be in order of ascending type id.
- */
 public class RelationshipStage extends Stage
```
Please add javadoc similar to the other stages, they are great.

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporter.java:
```diff
@@ -198,21 +200,21 @@ public void doImport( Input input ) throws IOException
-                    relationshipConfig,
-                    relationships, nodeRelationshipCache, idMapper, badCollector, inputCache, neoStore );
-            executeStage( calculateDenseNodesStage );
+            RelationshipStage unlinkedRelationshipStage =
```
Please add `// Step ... ` comment for consistency with rest of comments. (Also, I think they are useful when navigating the code)

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipStage.java:
```diff
@@ -19,84 +19,53 @@
+        add( new AssignRelationshipIdBatchStep( control(), config, 0 ) );
         add( new RelationshipPreparationStep( control(), config, idMapper ) );
         add( new RelationshipRecordPreparationStep( control(), config, neoStore.getRelationshipTypeRepository() ) );
-        add( new RelationshipEncoderStep( control(), config, cache ) );
```
`RelationshipEncoderStep` is not used anymore, it can be removed.

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporter.java:
```diff
@@ -300,7 +301,6 @@ private void importRelationships( NodeRelationshipCache nodeRelationshipCache,
         // finally there will be one Node --> Relationship and Relationship --> Relationship stage linking
         // all sparse relationship chains together.
```
This comment seems a bit outdated, can we update it?

burqen reviewed (changes requested):
Got some comments, mostly about javadoc

tinwelint committed:
Updated javadocs and comments in and around ParallelBatchImporter

tinwelint commented:
@burqen updated
